### PR TITLE
Add ability to share and receive loyalty cards

### LINKED ISSUE #249: [Feature Request] Share one or more entries
It would be good to be able to share one or more entries using standard share intents, in a way that would allow a contact to import them into their own locker.

### PR DESCRIPTION
**Pull request by TheLastProject**
This adds the ability to share and receive loyalty cards from an url.

Creating this pull request before the tests are in place to allow for early feedback.

Fixes #249 

TODO:
- [x] Write tests
- [x] Wait for nicer Url that in a browser will explain to the user they need to install the app (willing to help make the page, but I can't create the website itself)

photo_2019-12-01_18-24-35
photo_2019-12-01_18-24-36
photo_2019-12-01_18-24-38
photo_2019-12-01_18-24-39
photo_2019-12-01_18-24-40
photo_2019-12-01_18-24-41
photo_2019-12-01_18-24-43

---
This change is [Reviewable](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321)

### CODE REVIEW

**Comment by TheLastProject** [comment]
I would think the current tests may be sufficient, but please tell me if there are other parts of this you want me to write tests for.

**Comment by brarcher** [comment]
Oh, this is cool. I'd not thought of sharing a card working like this.

> Wait for nicer Url that in a browser will explain to the user they need to install the app (willing to help make the page, but I can't create the website itself)

I think that configuring a website is pretty easy with GitHub. For example, there could be a docs folder that contains the webpage:

Screen Shot 2019-12-01 at 6 14 17 PM

Details [here](https://pages.github.com/). If that was done, the page would be at:

`https://brarcher.github.io/loyalty-card-locker`

If you are interested in helping make a page using Jekyll, I can enable the option.

File: app/src/main/java/protect/card_locker/ImportURIHelper.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        try {
+            String store = uri.getQueryParameter(STORE);
+            String note = uri.getQueryParameter(NOTE);
+            String cardId = uri.getQueryParameter(CARD_ID);
+            String barcodeType = uri.getQueryParameter(BARCODE_TYPE);
+            Integer headerColor = Integer.parseInt(uri.getQueryParameter(HEADER_COLOR));
+            Integer headerTextColor = Integer.parseInt(uri.getQueryParameter(HEADER_TEXT_COLOR));
+            return new LoyaltyCard(-1, store, note, cardId, barcodeType, headerColor, headerTextColor);
+        } catch (NullPointerException | NumberFormatException ex) {
+            ex.printStackTrace();
```
Comment (brarcher) [inline app/src/main/java/protect/card_locker/ImportURIHelper.java]: Can remove this line

File: app/src/main/java/protect/card_locker/ImportURIHelper.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+public class ImportURIHelper {
+    private static final String STORE = DBHelper.LoyaltyCardDbIds.STORE;
+    private static final String NOTE = DBHelper.LoyaltyCardDbIds.NOTE;
+    private static final String CARD_ID = DBHelper.LoyaltyCardDbIds.CARD_ID;
+    private static final String BARCODE_TYPE = DBHelper.LoyaltyCardDbIds.BARCODE_TYPE;
+    private static final String HEADER_COLOR = DBHelper.LoyaltyCardDbIds.HEADER_COLOR;
+    private static final String HEADER_TEXT_COLOR = DBHelper.LoyaltyCardDbIds.HEADER_TEXT_COLOR;
+
+    private Context context;
```
Comment (brarcher) [inline app/src/main/java/protect/card_locker/ImportURIHelper.java]: These can be final

**Comment by TheLastProject** [comment]
I've never used GitHub Pages with the docs option so I'll have to take a look at what it expects me to put where, but sure, I'll give it a shot to create a very basic site for it

File: app/src/main/java/protect/card_locker/LoyaltyCardViewActivity.java [hunk 3]
```diff
@@ -221,6 +224,10 @@ public boolean onOptionsItemSelected(MenuItem item)
                 finish();
                 break;
 
+            case R.id.action_share:
+                importURIHelper.startShareIntent(importURIHelper.toUri(loyaltyCard));
```
Comment (brarcher) [inline app/src/main/java/protect/card_locker/LoyaltyCardViewActivity.java]: The two places that toUri is used it is passed directly to startShareIntent. Why not have startShareIntent take the card directly, and not expose toUri?

**Review by brarcher — CHANGES_REQUESTED** [review]
> Reviewed 10 of 12 files at r1, 3 of 4 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321)* status: all files reviewed, 7 unresolved discussions (waiting on @TheLastProject)

___
*[app/src/main/java/protect/card_locker/ImportURIHelper.java, line 34 at r3](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321#-Lv4Tw-jDOyarqSowsKN:-Lv4Tw-jDOyarqSowsKO:b-k675cn) ([raw file](https://github.com/brarcher/loyalty-card-locker/blob/2ef739fb4f9005c9200fea6aad134ff8b2634f7e/app/src/main/java/protect/card_locker/ImportURIHelper.java#L34)):*
> ```Java
>     }
> 
>     public LoyaltyCard parse(Uri uri) throws InvalidObjectException, InvalidParameterException {
> ```

What is the InvalidParameterException from? I do not see it being caught anywhere.
___
*[app/src/test/java/protect/card_locker/ImportURITest.java, line 21 at r3](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321#-Lv4V6Gn1eislgEYn_qE:-Lv4V6Gn1eislgEYn_qF:bgca3aj) ([raw file](https://github.com/brarcher/loyalty-card-locker/blob/2ef739fb4f9005c9200fea6aad134ff8b2634f7e/app/src/test/java/protect/card_locker/ImportURITest.java#L21)):*
> ```Java
> 
> import static org.junit.Assert.assertEquals;
> ```

Why the extra space between some of the imports?
___
*[app/src/test/java/protect/card_locker/ImportURITest.java, line 23 at r3](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321#-Lv4VRea1IkbM3AC0e7j:-Lv4VReb0a9lkv-013xz:b-i9bmrl) ([raw file](https://github.com/brarcher/loyalty-card-locker/blob/2ef739fb4f9005c9200fea6aad134ff8b2634f7e/app/src/test/java/protect/card_locker/ImportURITest.java#L23)):*
> ```Java
> 
> import static org.junit.Assert.assertTrue;
> import static protect.card_locker.DBHelper.LoyaltyCardDbIds.BARCODE_TYPE;
> ```

Having some uses of values from other classes is done using the class (e.g. Color.BLACK) and some are from a static import (BARCODE_TYPE). Can the static import of the value be changed to importing LoyaltyCardDbIds? The BARCODE_TYPE usage keep me thinking that the value is defined in this class, but it is not.
___
*[app/src/test/java/protect/card_locker/ImportURITest.java, line 38 at r3](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321#-Lv4W4B3EmsvrLz35xjv:-Lv4W4B3EmsvrLz35xjw:b-4dnsde) ([raw file](https://github.com/brarcher/loyalty-card-locker/blob/2ef739fb4f9005c9200fea6aad134ff8b2634f7e/app/src/test/java/protect/card_locker/ImportURITest.java#L38)):*
> ```Java
>         db = new DBHelper(activity);
>     }
> ```

Should there be a test for parsing a Uri in this class?

**Review by TheLastProject — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321)* status: 10 of 14 files reviewed, 7 unresolved discussions (waiting on @brarcher)

___
*[app/src/main/java/protect/card_locker/ImportURIHelper.java, line 18 at r3](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321#-Lv4TaA97xhCUInejI_t-r3-18:-Lv5TQwq07lixOPevV1F:b-896fix) ([raw file](https://github.com/brarcher/loyalty-card-locker/blob/2ef739fb4f9005c9200fea6aad134ff8b2634f7e/app/src/main/java/protect/card_locker/ImportURIHelper.java#L18)):*
<details><summary><i>Previously, brarcher (Branden Archer) wrote…</i></summary><blockquote>

These can be final
</blockquote></details>

Done.
___
*[app/src/main/java/protect/card_locker/ImportURIHelper.java, line 34 at r3](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321#-Lv4Tw-jDOyarqSowsKN:-Lv5Td_H0TFMJfdcHiqQ:b-aac5tq) ([raw file](https://github.com/brarcher/loyalty-card-locker/blob/2ef739fb4f9005c9200fea6aad134ff8b2634f7e/app/src/main/java/protect/card_locker/ImportURIHelper.java#L34)):*
<details><summary><i>Previously, brarcher (Branden Archer) wrote…</i></summary><blockquote>

What is the InvalidParameterException from? I do not see it being caught anywhere.
</blockquote></details>

Leftover, done now.
___
*[app/src/main/java/protect/card_locker/ImportURIHelper.java, line 48 at r3](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321#-Lv4TaA97xhCUInejI_t-r3-48:-Lv5TiJN167Up4nYXbTR:b-896fix) ([raw file](https://github.com/brarcher/loyalty-card-locker/blob/2ef739fb4f9005c9200fea6aad134ff8b2634f7e/app/src/main/java/protect/card_locker/ImportURIHelper.java#L48)):*
<details><summary><i>Previously, brarcher (Branden Archer) wrote…</i></summary><blockquote>

Can remove this line
</blockquote></details>

Done.
___
*[app/src/main/java/protect/card_locker/LoyaltyCardViewActivity.java, line 228 at r1](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321#-Lv4TaA97xhCUInejI_o-r1-228:-Lv5Tl-J7p2xuGwDSjON:b-896fix) ([raw file](https://github.com/brarcher/loyalty-card-locker/blob/28b3aa2018951a440de5649946562cdc2922e455/app/src/main/java/protect/card_locker/LoyaltyCardViewActivity.java#L228)):*
<details><summary><i>Previously, brarcher (Branden Archer) wrote…</i></summary><blockquote>

The two places that toUri is used it is passed directly to startShareIntent. Why not have startShareIntent take the card directly, and not expose toUri?
</blockquote></details>

Done.
___
*[app/src/test/java/protect/card_locker/ImportURITest.java, line 21 at r3](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321#-Lv4V6Gn1eislgEYn_qE:-Lv5Topd08zxqxcDlr1l:b-896fix) ([raw file](https://github.com/brarcher/loyalty-card-locker/blob/2ef739fb4f9005c9200fea6aad134ff8b2634f7e/app/src/test/java/protect/card_locker/ImportURITest.java#L21)):*
<details><summary><i>Previously, brarcher (Branden Archer) wrote…</i></summary><blockquote>

Why the extra space between some of the imports?
</blockquote></details>

Done.
___
*[app/src/test/java/protect/card_locker/ImportURITest.java, line 23 at r3](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321#-Lv4VRea1IkbM3AC0e7j:-Lv5TpYM0m5p30vQEKjP:b-896fix) ([raw file](https://github.com/brarcher/loyalty-card-locker/blob/2ef739fb4f9005c9200fea6aad134ff8b2634f7e/app/src/test/java/protect/card_locker/ImportURITest.java#L23)):*
<details><summary><i>Previously, brarcher (Branden Archer) wrote…</i></summary><blockquote>

Having some uses of values from other classes is done using the class (e.g. Color.BLACK) and some are from a static import (BARCODE_TYPE). Can the static import of the value be changed to importing LoyaltyCardDbIds? The BARCODE_TYPE usage keep me thinking that the value is defined in this class, but it is not.
</blockquote></details>

Done.
___
*[app/src/test/java/protect/card_locker/ImportURITest.java, line 38 at r3](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321#-Lv4W4B3EmsvrLz35xjv:-Lv5TqKsAM8lxMYg35Ih:buhlniv) ([raw file](https://github.com/brarcher/loyalty-card-locker/blob/2ef739fb4f9005c9200fea6aad134ff8b2634f7e/app/src/test/java/protect/card_locker/ImportURITest.java#L38)):*
<details><summary><i>Previously, brarcher (Branden Archer) wrote…</i></summary><blockquote>

Should there be a test for parsing a Uri in this class?
</blockquote></details>

I have an import with a valid URI in LoyaltyCardViewActivityTest (function importCard) indirectly because it deals with the intent and parses it there, so it seems duplicate to also put it here. Possible though.

**Comment by TheLastProject** [comment]
I believe this should be perfect now, assuming the URL will indeed become https://brarcher.github.io/loyalty-card-locker.

**Review by brarcher — APPROVED** [review]
> Reviewed 4 of 4 files at r4, 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/brarcher/loyalty-card-locker/321)* status: :shipit: complete! all files reviewed, all discussions resolved

**Comment by brarcher** [comment]
I've turned on GitHub Pages for the site, to serve from /docs from the master branch. It is not loading yet, but I'm anticipating that after some delay it will start serving. I'll check back tomorrow to see if it is working.

**Comment by TheLastProject** [comment]
It seems to not like my symlink from index.md in docs to README.md in the root, guess it doesn't have access to the root directory: https://github.com/brarcher/loyalty-card-locker/runs/330643870

I'll try to think of something soon, kinda don't want to make a copy of README.md if it can be prevented, but we may have to...

**Comment by TheLastProject** [comment]
https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-readmes

Ah, we can simply move the README.md into docs and then symlink index.md to README.md in the docs directory. It'll still be shown on the repo page. I'll make a pull request for that change tonight

**Review by pejfat08-rgb — APPROVED** [review]

**Pull request merged**